src/historywork/VerifyTxResultsWork.cpp

```diff
@@ -0,0 +1,157 @@
+#include "main/ErrorMessages.h"
+#include "util/FileSystemException.h"
+#include "util/Fs.h"
+#include "util/Logging.h"
+#include "util/format.h"
+
+namespace stellar
+{
+
+VerifyTxResultsWork::VerifyTxResultsWork(Application& app,
+                                         TmpDir const& downloadDir,
+                                         uint32_t checkpoint)
+    : BasicWork(app, "verify-results-" + std::to_string(checkpoint),
+                RETRY_NEVER)
+    , mDownloadDir(downloadDir)
+    , mCheckpoint(checkpoint)
+{
+}
+
+void
+VerifyTxResultsWork::onReset()
+{
+    mHdrIn.close();
+    mResIn.close();
+    mTxResultEntry = {};
+    mDone = false;
+    mEc = asio::error_code();
+}
+
+BasicWork::State
+VerifyTxResultsWork::onRun()
+{
+    if (mDone)
+    {
+        return mEc ? State::WORK_FAILURE : State::WORK_SUCCESS;
+    }
+
+    std::weak_ptr<VerifyTxResultsWork> weak(
+        std::static_pointer_cast<VerifyTxResultsWork>(shared_from_this()));
+    auto verify = [ weak, checkpoint = mCheckpoint ]()
+    {
+        auto self = weak.lock();
+        if (self)
+        {
+            asio::error_code ec;
+            auto verified = self->verifyTxResultsOfCheckpoint();
+            CLOG(TRACE, "History")
+                << "Transaction results verification for checkpoint "
+                << checkpoint
+                << (verified ? " successful"
+                             : (" failed: " +
+                                std::string(POSSIBLY_CORRUPTED_HISTORY)));
+            if (!verified)
+            {
+                ec = std::make_error_code(std::errc::io_error);
+            }
+
+            // Wake up call is posted back on the main thread
+            self->mApp.postOnMainThread(
+                [weak, ec]() {
+                    auto self = weak.lock();
+                    if (self)
+                    {
+                        self->mEc = ec;
+                        self->mDone = true;
+                        self->wakeUp();
+                    }
+                },
+                "VerifyTxResults: finish");
+        }
+    };
+
+    mApp.postOnBackgroundThread(verify, "VerifyTxResults: start in background");
+    return State::WORK_WAITING;
+}
+
+bool
+VerifyTxResultsWork::verifyTxResultsOfCheckpoint()
+{
+    try
+    {
+        FileTransferInfo hi(mDownloadDir, HISTORY_FILE_TYPE_LEDGER,
+                            mCheckpoint);
+        FileTransferInfo ri(mDownloadDir, HISTORY_FILE_TYPE_RESULTS,
+                            mCheckpoint);
+        mHdrIn.open(hi.localPath_nogz());
+        mResIn.open(ri.localPath_nogz());
+    }
+    catch (FileSystemException&)
+    {
+        CLOG(ERROR, "History") << "Transaction results failed verification: "
+                                  "are .xdr files downloaded?";
+        CLOG(ERROR, "History") << POSSIBLY_CORRUPTED_LOCAL_FS;
+        return false;
+    }
+
+    LedgerHeaderHistoryEntry curr;
+    while (mHdrIn && mHdrIn.readOne(curr))
```

Comment:
not sure about the way the code is organized:
we're supposed to iterate over `mResIn` and `mHdrIn` in lockstep, but the code is written as if this was not the case.
I think that `getCurrentTxResultSet` should instead be a function that ensures that we have the right result (right now the code also reads, which causes the code to be strange)